### DISCUSSION TITLE
error.js removes any possibility to customize the error message

### DISCUSSION THREAD

**Discussion by raphaelbadia** [vercel/next.js#49506] [Ideas]
### Describe the feature you'd like to request

When using classic error boundaries (like the [react-error-boundary](https://www.npmjs.com/package/react-error-boundary) package, we have access to the error, which allow us to customize the error message.

For instance let's imagine I have a page /posts/:id. 

In my server component I would like to do (very contrived example):

```jsx
const Page = async ({
  params: { postId },
}) => {
  // if (!postId) throw new Error('missing_post_id');
  
  const post = await getPost(id);
  if (!post) throw new Error('post_not_found');
  
  const author = await getPostAuthor(id);
  if (!author) throw new Error('post_has_no_author');
............
```

Then retrieve it in the error.js :

```
const Error = ({error}) => (
   <p>{error.message === 'post_not_found' ? 'The post doesnt exist': 'We couldnt find the post author'}</p>
)
```

However doing that results in the error being transformed into : 
> Error: An error occurred in the Server Components render. The specific message is omitted in production builds to avoid leaking sensitive details. A digest property is included on this error instance which may provide additional details about the nature of the error.
 
This makes us unable to explain errors to users, resulting in bad UX.

### Describe the solution you'd like

I would like to be able to opt out of the error behavior, and be able to get the error message.

### Describe alternatives you've considered

Maybe we could workaround that by splitting the code in multiple server component and putting a custom error.js for each component, but this would be a hassle !

**Comment by Fredkiss3** [comment]
The `error.js` is mostly for unexpected errors (500 or other), for 404 errors you can use a `not-found.js` file and you invoke it with the function `notFound()` imported from `next/navigation`, like so : 

```tsx

import { notFound } from 'next/navigation';

const Page = async ({
  params: { postId },
}) => {
  const post = await getPost(id);

  if (!post) {
    notFound();
  }

  // ... rest of your page
}

```

As for the first condition `if (!postId) throw new Error('missing_post_id');` , you don't need it because if your page is a [dynamic route segment](https://beta.nextjs.org/docs/routing/defining-routes#dynamic-segments) if you call your page with no id, it will route to the parent route if it exists ( eg: `/posts` & `/posts/:id` exist) else nextjs will show an 404 generic error page (the default one).

You use `error.js` when you encounter unexpected errors that happened during the rendering of your page (network failure during fetching for ex).

Docs for `not-found.js` is here : https://beta.nextjs.org/docs/api-reference/file-conventions/not-found

**Comment by raphaelbadia** [comment]
Thanks but that is not really what I wanted to outline. I have updated my example. Like you said, 

> The error.js is mostly for unexpected errors (500 or other), for 404 errors you can use a not-found.js

What about expected errors that are not 404 ? For instance, if I get a 401 because the auth expired ? Or if I get a 403 because I don't have access to the resource ? I need to write different error messages in these cases. 

I thought error.js would be my go-to in that situation...

**Comment by Fredkiss3** [comment]
For unauthorized errors, instead of throwing an error you would have to redirect to the login page (for ex), something like this :

```tsx

import { redirect } from 'next/navigation';
import { cookies } from 'next/headers';

const SecretLayout = async ({children}: { children: React.ReactNode }) => {
  const user = await getUser(cookies.get('session_token'));

  if (!user) {
    redirect(`/login`);
  }

  return <>{children}</>
}
```
I prefer to check for the user in a layout because that Layout can wrap a big part of the ui (a `DashboardLayout` for example ). as i said `error.js` is mostly for errors that you can't control or expect, to give users a way to recover instead of your app totally crashing.

**Comment by raphaelbadia** [comment]
Checking the user in the layout is a great idea, but how about the case where you consume an external API that uses JWT to authenticate ? (It's always my case)

Let's say we do that in a dashboard, the layout will check for the user session once when the page first load. But your JWT could expire 30 minutes later while you're clicking deep in your UI, leading to a 401 server side ! It was easy to handle that with good old error boundaries, but not anymore with this error.js 😭

Same scenario for 403s : you're in a dashboard, you go to a specific page, choose a specific feature... oh, you don't have the rights to do that. What happens ? To me, that is an *expected* error that I want to handle. If `error.js` isn't the place to do that, do you have any advice on how I should do it ? 🙏

**Comment by Fredkiss3** [comment]
For pages with 403 forbidden you can create another layout where you check if the user can do the action and maybe render an error instead of the page content.


```tsx
import { cookies } from 'next/headers';
import Link from 'next/link';

const SuperAdminLayout = async ({children}: { children: React.ReactNode }) => {
  const user = await getUser(cookies.get('session_token'));
  const hasAccess = await checkUserAccess(user);

  if (!hasAcces) {
    return ( 
      <div>
         <h1> Error 403 Forbidden, you do not have access to this page</h1>
         <Link href={`/dashboard`}>Return to dashboard</Link>
      </div>
     ) 
  }

  return <>{children}</>
}
```
PS: nextjs said that in the future there will be support for conditionnal routes that could make this easier : https://beta.nextjs.org/docs/routing/fundamentals#advanced-routing-patterns 


For the other case, i would suggest to check in the pages directly if the user is defined, and maybe create a utility function that automatically redirect to the login page for example : 

```tsx
// lib/functions.ts
import { cookies } from 'next/headers';

export async function getUserOrRedirect() {
  const user = await getUser(cookies.get('session_token'));
  
  if (!user) {
    redirect(`/login`);
  }

  return user;
}

// app/dashboard/protected/page.tsx
import { cookies } from 'next/headers';

const SecretPage = async () => {
  const user = await getUserOrRedirect();

  // ... your page content
}
```

And for further optimization if you fetch your user at many places in your layouts and server-components in one request, you can use `cache` to make it so that it fetches your user on the server only one per request : 


```ts
import { cache } from 'react';

export const getUser = cache((session_token: string) => {
  // fetch for the api or get the from the DB...
  return user;
});
```

**Comment by hnykda** [comment]
I stumbled upon this while trying to "propagate" HTTP Status code through error.js, as right now, it simply always returns HTTP status 200 (I haven't found a way how to make it return e.g. 500). The answers here do not address the issue, but rather recommend work around.

FIY, I consider handling this as the very basic property of a web framework, being used to having this in all frameworks starting back in 2005, but what do I know...

**Comment by raphaelbadia** [comment]
@hnykda I am unhappy too with the workarounds that work for specific use case but when dealing with an external API that has a lot of error case (stripe)

please upvote the issue if you agree :D

**Comment by Alxandr** [comment]
I have a similar issue where all requests to our backend includes a error code that needs to be displayed in the UI if the request fails. My current idea is to put this into a shortlived cookie or similar and then read it on the client, but that's definitely a hacky solution.

I would really like being able to return a custom error type with data that will be available to the error component, something like this:

```ts
import { SerializableError } from 'next';

export default function FailingComponent() {
  throw new SerializableError('error message', {
    // this here is an extra data object that gets json serialized and provided to the error component
    errorCode: 15
  });
}
```

If the error thrown is a `SerializableError`, the second argument to the constructor gets serialized and provided to the error component to do with as it likes. This makes using this opt in, and stack traces etc are still not sent to the client.

**Reply by willianrod** [reply]
Yep, I have the same situation here. I need to show custom error messages but there is no way to do that now

**Comment by Myks92** [comment]
I completely agree with the discussion. NextJS has no options to handle the 403 error from the api server. It's very sad. I hope it will be fixed!

**Comment by benokif** [comment]
I agree. The solutions showed by Fredkiss are good workarounds, but:
- they imply handling errors on each page (instead of having everything handled in a common `error.js`)
- they only work for GET requests. It is impossible to propagate errors from POST,PATCH or DELETE requests made from server actions in a Server component.

Take for instance a simple form in a server component, using a server action performing a `POST` request to a custom server.
A simple way to handle errors is to throw an `Error` based on the `response.status` of the `fetch()` request.
This is necessary, because there can be different errors at the time of the request (unauthorized, rate-limit hit), and we want to taylor the `error.js` displayed to the user based on that error.
This works nicely in development, but not in production because of [NextJS built-in sensitive information protection measure](https://nextjs.org/docs/app/building-your-application/routing/error-handling#securing-sensitive-error-information).

It works using Client components and `useFormState` as [described in the docs](https://nextjs.org/docs/app/building-your-application/data-fetching/forms-and-mutations#error-handling), but it would be nice to be able to use Server components too.
 
Can you allow us to opt-out of this configuration? For instance putting a `dangerouslyExposeServerError=true` in the config file.

**Comment by lu-zen** [comment]
Remix has support for caught and uncaught errors on the same function, Next DX is terrible on this matter and the docs lies about all errors being caught by error.js

**Comment by kamigerami** [comment]
In the context of handling errors from server actions, particularly with Prisma (e.g., unique constraint errors), how are others managing to relay informative yet secure error messages to users in a production environment? Given the security measures in Next.js that obscure error details, I'm looking for effective ways to communicate specific errors like these to users, possibly through UI notifications like toasts. Any insights or practices adopted in your projects would be greatly appreciated.

**Comment by benjick** [comment]
It would be nice to be able to `throw new PublicError("foo")` or similar and have it not be stripped of all data

**Comment by nicolo-tito** [comment]
Generally speaking, is someone already able to use the digest code in production in a sensible way? Like forwarding the server error and the digest code to an actual logger or an external service?

From what I see, the only place where the digest code appears (other than forwarded to error.js) is the stdout. I know there are ways to collect the stdout logging with third party libraries/services, but that's hacky at best and there's the need to manage it in a more straightforward way.

I don't know if discussions/solutions (maybe [this](https://nextjs.org/docs/app/api-reference/next-config-js/logging) in the future) are already in place, but I struggle to find info about this, I just find workarounds about using next-logger or experimental instrumentation hooks (basically monkey patching the console https://github.com/vercel/next.js/discussions/54719).

Very basic use case: I currently have to deploy my app on azure. On error, I would love to just get the actual server error and the digest code and send it to azure application insight (or a pino logger transport, or anything that I can manage in a centralized way), for it to be retrieved at a later time, using the digest code forwarded to the client.

(even sentry is struggling with this https://github.com/vercel/next.js/discussions/60283)

**Comment by hanayashiki** [comment]
It was 2023 and now 2024 and next.js doesn't have us handle errors in a typesafe way

**Reply by Abdelhammied** [reply]
I moved from next to Remix Due to this Senario

**Comment by HBLAB-TruongTT** [comment]
I'm facing this issue also

**Comment by renanbandeira** [comment]
I've been struggling with this and didn't want to add a code in all pages we have or add a props.error to the client component and handle it there as I can see in some articles like [this](https://blog.logrocket.com/testing-error-handling-patterns-next-js/#types-errors-next-js) one.

So I decided to have a wrapper for my pages, here it how it looked like (The ErrorPage basically gets the message and throws it to the user):
```js
type ServerComponentFn = ({ params, searchParams }) => Promise<JSX.Element>

export const withErrorCatch =
  (loadComponent: ServerComponentFn) =>
  async ({ params, searchParams }) => {
    try {
      const serverComponent = await loadComponent({ params, searchParams })
      return serverComponent
    } catch (error) {
      Sentry.captureException(error)
      // we should check if error.digest contains NEXT_REDIRECT as redirect throws an error
      if (isNextRedirectError(error as object)) {
        throw error
      }
      return <ErrorPage error={parseError(error as object)} />
    }
  }
```

and then, on my page.ts, instead of exporting the default page I would just do:
```js
export default withErrorCatch(Page)
```

It's not the ideal (I'd rather having a configuration to remove the server omit warning - although I understand the need of this omission), but it's the closest to the ideal I could think of

**Reply by 413n** [reply]
How did you handle the sync components?
Also, did you wrapped the layout.tsx too?

**Reply by renanbandeira** [reply]
I didn't wrap the layout file and I don't have sync server components in my codebase

**Reply by skolodyazhnyy** [reply]
You can make this middleware a bit more generic and use it for layouts too

```typescript
import { ReactElement } from 'react';

type ServerComponent<T> = (props: T) => Promise<ReactElement>;

const isError = (e: unknown): e is Error => e instanceof Error;
const isNextRedirectError = (e: unknown) => isError(e) && e.message === 'NEXT_REDIRECT';
const isNextNotFoundError = (e: unknown) => isError(e) && e.message === 'NEXT_NOT_FOUND';

export function withErrorCatch<T>(loadComponent: ServerComponent<T>) {
  return async (props: T) => {
    try {
      return await loadComponent(props);
    } catch (err) {
      Sentry.captureException(err)

      // we should check if error.digest contains NEXT_REDIRECT as redirect throws an error
      if (isNextRedirectError(err) || isNextNotFoundError(err)) {
        throw err;
      }

      return <ErrorPage error={isError(err) ? err : new Error(err + '')} />;
    }
  };
}

```

**Comment by ApacheEx** [comment]
I've managed to handle it in the following way (this is not ideal, but it works):


we need to have helper function (i.e wrapper):
```ts
// error-handler.ts
'use server';

import { isDynamicUsageError } from 'next/dist/export/helpers/is-dynamic-usage-error';

export const withErrorHandler = <T extends Array<unknown>, U>(
  fn: (...args: T) => Promise<U>,
) => {
  return async (...args: T): Promise<U | CustomError> => {
    try {
      return await fn(...args);
    } catch (e) {
      if (isDynamicUsageError(e)) throw e;

      return { error: { message: (e as Error).message } };
    }
  };
};

```
then you should wrap each server action with it, like this:
```ts
// server/actions.ts
'use server';

export const createFolder = withErrorHandler(
  async (params: RequestDto) => {
    /// your backend logic, which could potentially throw an exception.
  },
);
```

last but not least, you need to have some utils to catch this error on RSC:

```ts
// utils.ts
export type CustomError = { error: { message: string } };

export const isApiError = <T>(
  response: T | CustomError,
): response is CustomError => {
  return (
    typeof response === 'object' && response !== null && 'error' in response
  );
};

export function throwIfError<T>(
  response: T,
): asserts response is Exclude<T, CustomError> {
  if (isApiError(response)) {
    throw new Error(response.error.message);
  }
}

export function handleServerAction<T>(response: T): Exclude<T, CustomError> {
  throwIfError(response);

  return response;
}
```

finally, this is how you call server action in startTransition or RSC

```ts

 try {
  handleServerAction(await createFolder(values));
 } catch (e) {
  showError(e); // e.g shows toast based on e.message.
}

**Comment by 413n** [comment]
+1. An improvement to the DX of error handling would be really appreciated.
It's pretty limiting the fact that we can't force to ~~leak~~ get the `error.message` in production so to handle different error messages.
Also, to handle the RSCs errors with server-side code you must wrap it in a HOC as @renanbandeira did. It would be good if `error.tsx` could be a RSC that then could return a Client component with the reset button or something.

**Comment by isoroka-plana** [comment]
In my case calling `notFound()` in the server component gets caught by `error.tsx` 🤣 So I had to something like this:
```
if (isNotFoundError(error)) {
  throw error;
}
```
But now I am not sure if it will work in production because I tried following the same approach by assigning a custom digest value. Well, it doesn't work, I still get a hashed digest in the end.
The DX is really bad not only because it's not possible to customize this behavior but also because it works different in production than in dev. It all works great locally and then you get surprises when the code gets deployed. Not like anybody reads those little notes buried deep in your documentation somewhere, it should be intuitive and straightforward!

**Comment by nickjanssen** [comment]
I don't understand this design decision. Let's say I have a SaaS app with a customer that has exceeded its usage for its plan.
In my server action called from the client, I would want to throw an error such as HasExceededQuotaError with custom metadata (e.g. how much quota remaining), but Next.js strips this away making my error unusable. Now I have to resort to arbitrary return types like `{ success: false, remainingQuota: 4 }` and I can no longer use catch() in my client app.

**Reply by benjick** [reply]
Look at next-safe-action, you can handle the errors in middleware

**Comment by vrunda-inexture** [comment]
Instead of throwing the error, try returning an object with the error message. I encountered the same issue, and this approach resolved it for me.

**Comment by nickjanssen** [comment]
Here's where the error gets modified: https://github.com/facebook/react/blob/a07f5a3db5deb5a429bf2617525b6e66dc777e8c/packages/react-client/src/ReactFlightClient.js#L1711
My workaround is to return an object on the server and then throw it on the client. Then you can still display your own messages.

```ts
===
Server
===
export type ServerActionError = {
  isServerActionError: true;
  type: string;
  message: string;
};

export const updateAccount = async (data: ConnectAccountValues) => {
  let result = connectAccountSchema.safeParse(data);

  let error: ServerActionError | null = null;  

  if (!result.success) {
    error = {
      isServerActionError: true,
      type: "parse-error",
      message: result.error.message,
    };

    return error;    
  }

  // Get session
  const session = await getServerSession(authOptions);

  // Get user
  if (!session?.user) {
    error = {
      isServerActionError: true,
      type: "not-signed-in",
      message: "Please log in",
    };

    return error;
  }
  ...
===
Client
===
  try {
    const result = await updateAccount(data);

    if (result && result.isServerActionError) {
      throw result;
    }
  } catch (e: any) {
    alert(e.message);
  } 
```

**Reply by benjick** [reply]
Why not just redirect them to a login page?

**Reply by nickjanssen** [reply]
> Why not just redirect them to a login page?

This is a server action, AFAIK you can't "redirect" someone that is doing an ajax request. Also there are redirects in place higher up the chain (in layout.ts) but I still need to do a check here to prevent TS warnings.

**Reply by benjick** [reply]
Sure you can, I just tried it. Also if it's being checked higher up in the chain, just throw normally, no need to all this extra work which is never gonna trigger

**Reply by nickjanssen** [reply]
Thank you, will try soon!

**Comment by AhmedBaset** [comment]
I've found a solution to this

First of all, it has to be clear that this behavior is a React thing not Next.js https://github.com/facebook/react/blob/b7e7f1a3fab87e8fc19e86a8088a9e0fe4710973/packages/react-server-dom-webpack/src/__tests__/ReactFlightDOMEdge-test.js#L980-L983

In production, React omits error messages to mitigate security risks. But it doesn't override the `digest`, digest for who doesn't know is an auto-generated hash of the error that could be used to match the corresponding error in server-side logs. If you provided a `digest`, neither React nor Next.js will override it. 

```tsx
// page.tsx
class CustomError extends Error {
  digest = "I'm a custom digest";

  constructor() {
    super("Custom Error");
    this.name = "CustomError";
  }
}

export const dynamic = "force-dynamic";

export default function Page() {
  if (new Date().getMinutes() >= 48) throw new CustomError();

  return "Null";
}
```

```tsx
// error.tsx

"use client";

export default function Error({
  error,
}: { error: Error & { digest: string } }) {
  return (
    <div dir="ltr">
      <p>Message: {error.message}</p>
      <p>Digest: "{error.digest}"</p>
    </div>
  );
}
```

![image](https://github.com/user-attachments/assets/036f09c8-5aa2-4b7c-9d50-aa2ebc967bf0)

Then, you can rely on the digest to handle your errors, for example

```tsx
// error.tsx
"use client";

export default function Error({
  error,
}: { error: Error & { digest: string } }) {
  switch (error.digest) {
    case "HTTP:400":
      return <p>Bad Request</p>;
    case "HTTP:401":
      return <p>Unauthorized</p>;
    case "HTTP:403":
      return <p>Forbidden</p>;
    // ...
  }

  return <p>Another error</p>;
}
```
```tsx
// page.tsx
export const dynamic = "force-dynamic";

export default function Page() {
  const error = new Error();
  (error as any).digest = "HTTP:401";

  if (new Date().getMinutes() >= 5) throw error;

  return "Null";
}
```

**Reply by jonshipman** [reply]
Hilariously fun bug to find. Would be nice if we could figure out a way to forward all error.message to error.digest without a huge refactor.

**Reply by raphaelbadia** [reply]
Wow, finally a solution ! Thank you for this 💫

**Reply by SarahVanDenBerghe** [reply]
This was just what I was looking for! I previously added route handlers in NextJS (app router) in order to catch and read the server errors, which could be passed through client with NextResponse, but it's a lot of set-up for just wanting to retrieve the status code. Thank you!

**Reply by HOSS11H** [reply]
i get digest as undefined on production build only v 14.1.0 . any idea why ?

**Comment by dealer-solutions-gene** [comment]
So I'm here because... of course I am after finding this behavior with custom error handling.

I needed more data on my error screen, so I've taken advantage of the digest being a `string` and cramming it with JSON..

    export class MyCustomError extends Error {
      public myMessage: string;
      public foo: string;
      private bar: string;
      constructor(message: string, foo: string, bar: string) {
        super(message);
        this.myMessage = message;
        this.foo = foo;
        this.bar = bar;
      }
      get digest(): string {
        return JSON.stringify(this);
      }
    }

The `get digest()` will contain whatever you want from your error class as JSON when `error.digest` is read. Wrap a `JSON.parse(error.digest)` in a try/catch if you need to, and omit the sensitive things if you need to etc etc

edit: an extra copy of the message (eg `myMessage`) might be needed depending on when message is stripped and digest is fetched by NextJS

----

Work in progress: but a "re-hydration" of the object in TypeScript (I'm not looking into needing it as the whole class right now). Build your own rules as required...

    export const decodeErrorWithDigestAsMyCustomError = (error: Error & { digest?: string }) => {
      let customError: undefined | MyCustomError = undefined;
      try {
        if (!!error.digest) {
            customError = JSON.parse(error.digest) as unknown as MyCustomError;
          if (isMyCustomErrorTypeGuardFunction(customError)) {
            return customError;
          }
        }
      } catch (e) {
        // JSON Parse error
      }
      return undefined;
    };
